### PR TITLE
Extend cluster itest

### PR DESCRIPTION
**Pull request by lucasponce**
This PR extend the ITests coverage on cluster profile.

There is a known issue about listeners that is being addressed on HWKALERTS-223 but that will be sent in a separate PR.

This PR have the cluster profile stable.

Please @jshaughn, take a look. Thx.

### CODE REVIEW

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java [hunk 1]
```diff
@@ -52,12 +52,13 @@ public AlertsEngineCache() {
...
      *
+     * @param tenantId to check if has triggers deployed on this node
      * @param dataId to check if it has triggers deployed on this node
      * @return true if it is active
      *         false otherwise
      */
-    public boolean isDataIdActive(String dataId) {
-        return dataId != null && activeDataIds.contains(dataId);
+    public boolean isDataIdActive(String tenantId, String dataId) {
+        return tenantId != null && dataId != null && activeDataIds.contains(new DataId(tenantId, dataId));
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java]: Was adding the tenant fixing a bug or added as some sort of formality?  If I recall, ignoring tenant was a conscious decision to avoid the extra object creation and to consolidate the cache entries. Was there an issue with remove being too aggressive?

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java [hunk 2]
```diff
@@ -467,22 +467,27 @@ private void removeTrigger(Trigger trigger) {
...
             rules.removeFacts(t -> {
                 if (t instanceof Dampening) {
-                    return ((Dampening) t).getTriggerId().equals(triggerId);
+                    return ((Dampening) t).getTenantId().equals(tenantId)
+                            && ((Dampening) t).getTriggerId().equals(triggerId);
                 } else if (t instanceof Condition) {
-                    return ((Condition) t).getTriggerId().equals(triggerId);
+                    return ((Condition) t).getTenantId().equals(tenantId) &&
+                            ((Condition) t).getTriggerId().equals(triggerId);
                 }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Wow, yes, this was definitely a problem. Good catch.  I still wonder whether we should do this in rules to simplify the java side.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java [hunk 3]
```diff
@@ -521,6 +526,9 @@ private void addData(TreeSet<Data> data) {
         }
 
         synchronized (pendingData) {
+            if (log.isDebugEnabled()) {
+                log.debugf("Adding [%s] to pendingData [%s]", data, pendingData);
+            }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: trivial: you can remove the isDebugEnabled wrapper

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java [hunk 4]
```diff
@@ -575,6 +583,9 @@ private void addEvents(TreeSet<Event> events) {
         }
 
         synchronized (pendingEvents) {
+            if (log.isDebugEnabled()) {
+                log.debugf("Adding [%s] to pendingEvents [%s]", events, pendingEvents);
+            }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Same, can remove the wrapper.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java [hunk 5]
```diff
@@ -762,19 +773,21 @@ private void handleAutoResolvedTriggers() {
 
     private int checkMissingStates() {
         int numMatchingEvals = 0;
-        for (MissingState missingState : missingStates) {
+        Iterator<MissingState> itMissingStates = missingStates.iterator();
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: It's fine, but why did you change this to use an Iterator?

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java [hunk 6]
```diff
@@ -270,6 +270,9 @@ public void notifyData(Collection<Data> data) {
         if (distributed) {
             NotifyData nData = new NotifyData(currentNode, data, Data.class);
             Integer key = nData.hashCode();
+            if (log.isDebugEnabled()) {
+                log.debugf("Sending data [%s]", nData);
+            }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: same, can remove wrapper.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java [hunk 7]
```diff
@@ -280,6 +283,9 @@ public void notifyEvents(Collection<Event> events) {
         if (distributed) {
             NotifyData nEvent = new NotifyData(currentNode, events, Event.class);
             Integer key = nEvent.hashCode();
+            if (log.isDebugEnabled()) {
+                log.debugf("Sending events [%s]", nEvent);
+            }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: sorry for the repetition, same...

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java [hunk 8]
```diff
@@ -740,6 +746,9 @@ private void processNotifyData(NotifyData notifyData) {
             if (!dataListeners.isEmpty() && notifyData.getFromNode() != currentNode) {
                 if (notifyData.getDataCollection() != null) {
                     dataListeners.stream().forEach(dataListener -> {
+                        if (log.isDebugEnabled()) {
+                            log.debugf("processNotifyData [%s]", notifyData);
+                        }
```
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: same.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java (hunk 1 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java]: If this function doesn't have a tenant filter, it might send wrong data/events to the core engine.
At the end, data/events are discarded but the engine itself, but this limits the flow.
Yes, it is a bug, it doesn't show up in the final result, but it makes the engine to process wrong data.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 2 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: I think rules side is safe.
For the moment, this fix is the only needed to fix the problem.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 3 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 4 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 5 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Because missingStates can be updated concurrently by other thread.
So using a for is not safe on this case.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java (hunk 6 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java (hunk 7 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java (hunk 8 above)
Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/PartitionManagerImpl.java]: +1

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java (hunk 1 above)
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineCache.java]: OK, so if in a distributed environment triggers with the same ids and different tenants can be processed on different nodes, then this makes sense.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 2 above)
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Perhaps someday I will create a PR for this to be rule based, I am fine with the change here, though.

File: hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java (hunk 5 above)
Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Are you sure Iterator will help this situation?  It's true that Iterator offers the ability to remove the current item but I'm not sure it is really thread safe.

Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: It's handled in two ways.
The updated part is synchronized by the missingStates, and the read part is moved to use an iterator.
So, the iterator itself doesn't solve, that's protected via the synchronized, but moving to an iterator, it check on each element, which then is safe on potential modifications.

Comment (jshaughn) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: I will believe you.  Although, I always thought the for-each construct was just syntactic sugar on top of an underlying Iterator, which would make this change a no-op.  Whatever is happening, it seems subtle in the code and may merit an inline comment about the concurrency issue and how this is safe.

Comment (lucasponce) [inline hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java]: Yes you are right.
The bytecode for each vs iterator is the same.
I first moved to an iterator, and when I discover the problem I protected the update with a synchronize in the other side.
Well, I would maintain it, at least for me it helps to remind that a collection managed by iterator can be updated outside of the loop concurrently.
